virttest/utils_libvirt/libvirt_ceph_utils.py

```diff
@@ -0,0 +1,154 @@
+    if ceph_client_name and ceph_client_key:
+        with open(key_file, 'w') as f:
+            f.write("[%s]\n\tkey = %s\n" %
+                    (ceph_client_name, ceph_client_key))
+        key_opt = "--keyring %s" % key_file
+    if is_setup:
+        # If enable auth, prepare device source
+        if ceph_client_name and ceph_client_key:
+            auth_sec_dict = {"sec_usage": auth_sec_usage_type,
+                             "sec_name": "ceph_auth_secret"}
+            auth_sec_uuid = libvirt.create_secret(auth_sec_dict)
+            virsh.secret_set_value(auth_sec_uuid, ceph_auth_key,
+                                   debug=True)
+            disk_auth_dict = {"auth_user": ceph_auth_user,
+                              "secret_type": auth_sec_usage_type,
+                              "secret_uuid": auth_sec_uuid}
+            device_source = "rbd:%s:mon_host=%s:keyring=%s" % (ceph_disk_name,
+                                                               ceph_mon_ip,
+                                                               key_file)
+        else:
+            device_source = "rbd:%s:mon_host=%s" % (ceph_disk_name, ceph_mon_ip)
+        logging.debug("device source is: %s", device_source)
+        # clean up image file if exists
+        cmd = ("rbd -m {0} {1} info {2} && rbd -m {0} {1} rm "
+               "{2}".format(ceph_mon_ip, key_opt, ceph_disk_name))
+        process.run(cmd, ignore_status=True, shell=True, verbose=True)
+
+        #Create necessary image file if not exists
+        if img_file is None:
+            img_file = os.path.join(data_dir.get_tmp_dir(),
+                                    "%s_test.img" % vm.name)
+            # Create an local image and make FS on it.
+            disk_cmd = ("qemu-img create -f %s %s %s" %
+                        (device_format, img_file, storage_size))
+            process.run(disk_cmd, ignore_status=False, shell=True)
+        # Convert the image to remote ceph storage
+        disk_path = ("rbd:%s:mon_host=%s" %
+                     (ceph_disk_name, ceph_mon_ip))
+        if ceph_client_name and ceph_client_key:
+            disk_path += (":id=%s:key=%s" %
+                          (ceph_auth_user, ceph_auth_key))
+        rbd_cmd = ("rbd -m %s %s info %s 2> /dev/null|| qemu-img convert -O"
+                   " %s %s %s" % (ceph_mon_ip, key_opt, ceph_disk_name,
+                                  device_format, img_file, disk_path))
+        process.run(rbd_cmd, ignore_status=False, shell=True, verbose=True)
+
+        # Disk related config
+        disk_src_dict = {"attrs": {"protocol": "rbd",
+                                   "name": ceph_disk_name},
+                         "hosts":  [{"name": ceph_mon_ip,
+                                     "port": ceph_host_port}]}
+        # Create disk xml with given config
+        vmxml = vm_xml.VMXML.new_from_dumpxml(vm.name)
+        disk_xml = Disk(type_name='network')
```

Comment:
Can you reuse libvirt.create_disk_xml() or libvirt.set_vm_disk() for this part?

Reply:
I don't think libvirt.create_disk_xml() or libvirt.set_vm_disk()  suit here since both of them involve in too many parameters which are not related, Here we only need one simple diskxml. For example create_disk_xml

            source_protocol = params.get("source_protocol")
            source_name = params.get("source_name")
            source_host_name = params.get("source_host_name").split()
            source_host_port = params.get("source_host_port").split()
            transport = params.get("transport")
            source_attrs = {'protocol': source_protocol, 'name': source_name}

Reply:
I think it support for any xml configuration, if you only need to give some of the parameters, it will generate xml with only what you are given. Please begin to use it as it is really useful and can make the code clean. It is our rule to write new cases that using the common function as much as possible. Thanks

Reply:
> I think it support for any xml configuration, if you only need to give some of the parameters, it will generate xml with only what you are given. Please begin to use it as it is really useful and can make the code clean. It is our rule to write new cases that using the common function as much as possible. Thanks

@kylazhang 
In principal, I totally agree we need call current utility as we can as possible.
Let's elaborate more additional burdens If I use create_disk_xml:
Firstly I need rewrappped below parameters since parameter name is not consistent:
```
    params.update({"type_name" : 'network'})
    params.update({"target_dev" : device_target})
    params.update({"target_bus", device_bus})
    params.update({"device_type", device})
     params.update({"ceph_user", xx})
  .   ...
     ceph-related
secret-related
```
Secondly, auth_attrs is used to diff auth in source or out source in create_disk_xml, so I need prepare for another paramter

Thirdly, since I use netwrok, I need prepare more parmeter
'''
            source_protocol = params.get("source_protocol")
            source_name = params.get("source_name")
            source_host_name = params.get("source_host_name").split()
            source_host_port = params.get("source_host_port").split()
            transport = params.get("transport")
            source_attrs = {'protocol': source_protocol, 'name': source_name}
'''

So my question, so I need pull so many ruduants works ,but actually only 6 lines  can work out in this PR.
Hopefully you can understand more now.

Reply:
> @kylazhang
> In principal, I totally agree we need call current utility as we can as possible.
> Let's elaborate more additional burdens If I use create_disk_xml:
> Firstly I need rewrappped below parameters since parameter name is not consistent:
> 
> ```
>     params.update({"type_name" : 'network'})
>     params.update({"target_dev" : device_target})
>     params.update({"target_bus", device_bus})
>     params.update({"device_type", device})
>      params.update({"ceph_user", xx})
>   .   ...
>      ceph-related
> secret-related
> ```

I think you can update them all with 1 line, or just use the same name as in create_disk_xml, then this line is also not needed and may reduce some params.get ...

> 
> Secondly, auth_attrs is used to diff auth in source or out source in create_disk_xml, so I need prepare for another paramter
> 
I saw there are auth_attrs setting in different place in this func.
```
            if auth_in_source:
                disk_source.auth = diskxml.new_auth(**auth_attrs)
            else:
                diskxml.auth = diskxml.new_auth(**auth_attrs)
```
And just 1 param in cfg with the same name "auth_in_source" and should be no update needed in py file, I think it should be acceptable

> Thirdly, since I use netwrok, I need prepare more parmeter
> '''
> source_protocol = params.get("source_protocol")
> source_name = params.get("source_name")
> source_host_name = params.get("source_host_name").split()
> source_host_port = params.get("source_host_port").split()
> transport = params.get("transport")
> source_attrs = {'protocol': source_protocol, 'name': source_name}
> '''
No, seems it just meet your need
```
source_attrs = {'protocol': source_protocol, 'name': source_name}
for host_name, host_port in list(
            zip(source_host_name, source_host_port)):
     source_host.append({'name': host_name,
                                        'port': host_port})

```
just give 'source_protocol', 'source_name','source_host_name','source_host_type' in params(update params or just setting in cfg file), and may no needed for your setting in code
```
        disk_src_dict = {"attrs": {"protocol": "rbd",
                                   "name": ceph_disk_name},
                         "hosts":  [{"name": ceph_mon_ip,
                                     "port": ceph_host_port}]}
```
> 
> So my question, so I need pull so many ruduants works ,but actually only 6 lines can work out in this PR.
> Hopefully you can understand more now.

Please correct me if I have some misunderstanding as I do not update disk too much.

And more, the lines count in code may not prove everything, and the most we want to implement is to make the code more clear, tidy, easy to update and maintain, and for common functions are more widely used. Thanks

Reply:
@kylazhang 
Quote "And more, the lines count in code may not prove everything, and the most we want to implement is to make the code more clear, tidy, easy to update and maintain, and for common functions are more widely used. Thanks"

I will invite you to join virtual disk cases automation. Look forward to your practice "And more, the lines count in code may not prove everything, and the most we want to implement is to make the code more clear, tidy, easy to update and maintain, and for common functions are more widely used. Thanks""

Reply:
In order to move this PR forward, I will attempt to create separate method to create disk.

Reply:
Then that's OK, if the purpose is to avoid the circle import issues by not using create_disk_xml, I can accept the current implement method here

Reply:
After discussion, please help create new function in virttest/utils_libvirt/libvirt_disk.py to create the function for creating a disk.